### PR TITLE
feat: a form's initial data can now be supplied as a dynamic callback

### PR DESCRIPTION
**Pull request by jlews**
## What's new?

# Initial data as a callback

Initial state can also be passed as a React callback. The callback will be executed:
1. Once initially with no parameters passed to it, it will be expected to return the initial state for the form binder.
2. Again each time it's dependencies update, in this case it will receive the current form state as a param and be expected to return updated form state.

This is extremely useful when a form relies on remote data that could update at any time, it allows the form to update with the latest changes from the server without losing the user's current edits by completely replacing the form data with the remote update.

**NOTE:** When using an initial data function, be aware that it will be executed to provide initial data whenever it's reference updates. It's essential therefore that any initial data function is a callback with a carefully selected set of dependencies, rather than an anonymous function or constant function.

Here's an example:
```tsx
// this is data from the server that can be updated at any time.
const { data } = useGetProfileData<IUser>();

// Using an initial data callback here ensures that when the remote data updates, the form will
// reflect the changes but also maintain live user edits to the uneffected areas of the form.
const initialData = React.useCallback((currentFormState?: IUser) => {
  return {
    firstName: "",
    lastName: "",
    email: ""
    ...(currentFormState ?? {}),
    ...(data ?? {})
  }
}, [data]);

const { formProp, formState } = Form.use(initialData);

const submitData = React.useCallback(async () => {
  await updateProfile(formState);
}, [formState]);

return (
  <form>
    <TextInput bind={formProp('firstName').bind()} />
    <TextInput bind={formProp('lastName').bind()} />
    <EmailInput bind={formProp('email').bind()} />
    <Button type="submit" onClick={submitData}>
      Update
    </Button>
  </form>
);
```

[board](https://rocketmakers.atlassian.net/jira/software/projects/HM/boards/172)

## Checklist

- [✅] are your changes in Storybook?
- [✅] are any breaking changes documented in `docs/migrating_from_oldstrong.md`?
- [✅] does _everything_ have jsdoc?
- [✅] is everything exported from index.ts?
- [✅] are all new hooks added to `src/hooks/hooks.stories.mdx` or given their own docs in Storybook?
- [✅] are all new SCSS mixins added to `src/theme/mixins.stories.mdx`?
- [✅] are all new SCSS variables added to `src/theme/variables.stories.mdx`?

### CODE REVIEW

File: playground/src/views/user/user.tsx [hunk 1]
```diff
@@ -1,4 +1,4 @@
-import * as React from "react";
```
Comment (jlews) [inline playground/src/views/user/user.tsx:1]: No need to review this file, playground changes

File: playground/src/servers/memory.ts [hunk 2]
```diff
@@ -20,13 +20,13 @@ export namespace MemoryServer {
     isCool?: boolean
```
Comment (jlews) [inline playground/src/servers/memory.ts:20]: No need to review this file, playground changes

**Review by adamiprinciples — APPROVED** [review]
Nice.

File: module/src/hooks/form/form.utils.ts [hunk 3]
```diff
@@ -77,3 +78,14 @@ export function isArrayValue(value: any, attemptedAction: string): value is any[
...
   return true;
 }
+
+/**
+ * Detects whether the incoming initial form data is a callback or an object and casts appropriately.
+ * @param initialData Either some initial form data, or a function that returns initial form data.
+ * @returns true if param is a function, also casts appropriately.
+ */
+export function initialDataIsCallback<TData extends object>(
+  initialData?: TData | InitialDataFunction<TData>
```
Comment (patrickbenjamin1) [inline module/src/hooks/form/form.utils.ts:88]: change this to theDataThatIsInitial

Comment (patrickbenjamin1) [reply module/src/hooks/form/form.utils.ts:88]: or preferably `dataOnlyItIsTheDataThatIsInitialAsInTheFirstData`

Comment (adamiprinciples) [reply module/src/hooks/form/form.utils.ts:88]: Resolving due to general clownery

**Pull request merged**